# Olist E-Commerce — Data Cleaning & Database Build

This notebook is **stage 1** of the project: it takes the 9 raw CSVs of the [Brazilian E-Commerce Public Dataset by Olist](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce) (~100K orders, 2016–2018) and turns them into a cleaned, explicitly-typed PostgreSQL database with enforced primary and foreign keys.

Downstream stages build on the database created here:

- **SQL analysis** — six business questions, logged with queries, findings and caveats in `Olist-Analysis-Log.md` (queries in `sql/`)
- **Power BI dashboard** — a 4-page report built on views exported from this database (`olist.pbix`)

| Step | What happens | Key decision |
|---|---|---|
| 0 | Load raw CSVs | — |
| 1 | Duplicate rows | Dedup reviews on `order_id`, not the documented-buggy `review_id` |
| 2 | Geolocation | Collapse ~1M noisy rows to one row per zip prefix |
| 3 | Missing values | Fill only where NULL is not informative; keep NULLs that carry meaning |
| 4 | Data types | Strict datetime parsing (`errors='raise'`), nullable `Int64` counts |
| 5 | Schema & load | Explicit SQLAlchemy schema with real PK/FK constraints |
| 6 | Integrity checks | Verify zero orphaned child rows |

## Step 0 — Load the raw data

All 9 tables load into a `dfs` dict (for loop-style audits across every table at once) plus individual named references (for per-table cleaning).


In [1]:
import pandas as pd
import numpy as np

dfs = {}

files = {
    "customers": "data/olist_customers_dataset.csv",
    "geolocation": "data/olist_geolocation_dataset.csv",
    "order_items": "data/olist_order_items_dataset.csv",
    "order_payments": "data/olist_order_payments_dataset.csv",
    "order_reviews": "data/olist_order_reviews_dataset.csv",
    "orders": "data/olist_orders_dataset.csv",
    "products": "data/olist_products_dataset.csv",
    "sellers": "data/olist_sellers_dataset.csv",
    "translation": "data/product_category_name_translation.csv"
}

for name, path in files.items():
    dfs[name] = pd.read_csv(path, on_bad_lines="warn", engine='python')


customers, geolocation, order_items = dfs["customers"], dfs["geolocation"], dfs['order_items']
order_payments, order_reviews, orders = dfs['order_payments'], dfs['order_reviews'], dfs['orders']
products, sellers, translation  = dfs['products'], dfs['sellers'], dfs['translation']

## Step 1 — Duplicate rows

First, a full-row duplicate audit across all 9 tables — before trusting any table, check whether the same record appears twice.


In [2]:
for name, df in dfs.items():
    duplicated = df.duplicated().sum()
    print(f"{name} Table have {duplicated} duplicated rows")

customers Table have 0 duplicated rows
geolocation Table have 261831 duplicated rows
order_items Table have 0 duplicated rows
order_payments Table have 0 duplicated rows
order_reviews Table have 0 duplicated rows
orders Table have 0 duplicated rows
products Table have 0 duplicated rows
sellers Table have 0 duplicated rows
translation Table have 0 duplicated rows


Only `geolocation` (261,831 full-row duplicates — handled in Step 2) is affected at the full-row level. But full-row checks aren't enough: `order_reviews` needs a **key-level** check, because the official data dictionary warns that `review_id` can repeat due to platform crashes and bugs.


In [3]:
# Checking duplication in order_reviews
# The data dictionary mentioned review_id can be repeated due to crashes and bugs
print(order_reviews['order_id'].duplicated().sum())
print(order_reviews['review_id'].duplicated().sum())

551
814


Both keys are dirty: 814 duplicated `review_id`s *and* 551 duplicated `order_id`s. Decision: deduplicate on **`order_id`** — the analyses that follow treat a review as an attribute of an order (one review per order), and `order_id` is the trustworthy key while `review_id` is the one documented as bug-prone.


In [4]:
# Deduping order_id as it is more trustworthy than review_id
order_reviews = order_reviews.drop_duplicates(subset='order_id', keep='first')
order_reviews['order_id'].duplicated().sum()

np.int64(0)

## Step 2 — Geolocation: one row per zip prefix

The largest and noisiest table in the dataset (~1M rows). Beyond the 261,831 exact duplicates, many rows differ only in accented vs. unaccented spellings of the same city (`são paulo` vs `sao paulo`). Normalizing to plain ASCII (NFKD decomposition, strip combining marks) exposes 17,837 additional duplicates that pure `drop_duplicates` missed.


In [5]:
geolocation.drop_duplicates(inplace=True)
# Cleaning the mixed accented letter by converting them to plain english letters
geolocation['geolocation_city'] = geolocation['geolocation_city'].str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')
geolocation.duplicated().sum()

np.int64(17837)

The goal is to use `zip_code_prefix` as a join key to `customers` and `sellers` — which requires it to be **unique**. It isn't yet: this check counts zip prefixes that still map to more than one city or state.


In [6]:
# Checking the amount of city with conflicted zip code
check = geolocation.groupby("geolocation_zip_code_prefix")[['geolocation_city', 'geolocation_state']].nunique().loc[lambda df: (df > 1).any(axis=1)]
check.shape

(555, 2)

555 zip prefixes (~3% of unique zips) carry conflicting city/state values. Decision: aggregate to exactly one row per zip prefix — **mean** lat/lng (a centroid for mapping) and **modal** city/state (the majority label wins). For 97% of zips nothing changes; for the conflicted 3%, the most frequent city is a defensible representative.


In [7]:
# Aggregating the lat lng, grouped by the zip code to make zip code unique, allowing us to use it as a foreign key to connect to the Customers and Sellers Table
# Since the zip code conflicted occurred only for about 3% of the data, so only picking the major city as representation. 
geolocation = geolocation.groupby('geolocation_zip_code_prefix').agg({
    'geolocation_lat': 'mean',
    'geolocation_lng': 'mean',
    'geolocation_city': lambda x: x.mode().iloc[0],
    'geolocation_state': lambda x: x.mode().iloc[0]
}).reset_index()
print(geolocation.head(5))
print(geolocation.shape)

   geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1001       -23.550227       -46.634039   
1                         1002       -23.547657       -46.634991   
2                         1003       -23.549000       -46.635582   
3                         1004       -23.549829       -46.634792   
4                         1005       -23.549547       -46.636406   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP  
(19015, 5)


Verification: rerunning the conflict check on the aggregated table returns an empty frame — every zip prefix now maps to exactly one city and state, so it can serve as a primary key.


In [8]:
geolocation.groupby("geolocation_zip_code_prefix")[['geolocation_city', 'geolocation_state']].nunique().loc[lambda df: (df > 1).any(axis=1)]

,geolocation_city,geolocation_state
geolocation_zip_code_prefix,,


## Step 3 — Missing values

A full audit of every table: which columns have NULLs, how many, and what share of the table they represent. The percentages drive the decisions — a 0.16% gap and an 88% gap call for different treatment.


In [9]:
for name, df in dfs.items():
    missing = df.isnull().sum()
    missing = missing[missing > 0]
    
    print(f"\n--- {name.upper()} ---")
    if missing.empty:
        print("No missing values")
    else:
        for col, count in missing.items():
            pct = (count / len(df)) * 100
            print(f"{col}: {count} ({pct:.2f}%)")


--- CUSTOMERS ---
No missing values

--- GEOLOCATION ---
No missing values

--- ORDER_ITEMS ---
No missing values

--- ORDER_PAYMENTS ---
No missing values

--- ORDER_REVIEWS ---
review_comment_title: 87656 (88.34%)
review_comment_message: 58247 (58.70%)

--- ORDERS ---
order_approved_at: 160 (0.16%)
order_delivered_carrier_date: 1783 (1.79%)
order_delivered_customer_date: 2965 (2.98%)

--- PRODUCTS ---
product_category_name: 610 (1.85%)
product_name_lenght: 610 (1.85%)
product_description_lenght: 610 (1.85%)
product_photos_qty: 610 (1.85%)
product_weight_g: 2 (0.01%)
product_length_cm: 2 (0.01%)
product_height_cm: 2 (0.01%)
product_width_cm: 2 (0.01%)

--- SELLERS ---
No missing values

--- TRANSLATION ---
No missing values


### Findings and decisions

**Already clean:** `customers`, `geolocation`, `order_items`, `order_payments`.

**`order_reviews`** — `review_comment_title` is 88.3% missing and `review_comment_message` 58.7% missing. That is expected behavior, not corruption: most buyers leave a star rating without writing a title or a message. Decision: fill with the explicit sentinels `'No Title'` / `'No Message'` so downstream SQL can distinguish "no comment" from a genuinely absent record.

**`orders`** — `order_approved_at`: 160 (0.16%), `order_delivered_carrier_date`: 1,783 (1.79%), `order_delivered_customer_date`: 2,965 (2.98%). These NULLs are informative: they mark orders that were canceled or never completed a given stage of the pipeline. Imputing a fake timestamp would fabricate deliveries that never happened, so they are deliberately **left as NULL** — later analyses (delivery status vs. review score) depend on exactly this signal.

**`products`** — 610 products (1.85%) have no category and 2 have no physical dimensions. They can't be dropped: their `product_id`s are referenced by real sales in `order_items`. Decision: fill the categorical column with `'Unknown'`, keep numeric dimensions as nullable — imputing means/zeros would silently distort category averages.


The two fill operations below implement the decisions above; each is verified with a fresh NULL count.


In [10]:
# Cleaning order_reviews table

order_reviews['review_comment_title'] = order_reviews['review_comment_title'].fillna('No Title')
order_reviews['review_comment_message'] = order_reviews['review_comment_message'].fillna('No Message')

print(order_reviews.isnull().sum())

review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64


In [11]:
# Cleaning Products table
""" 
Columns product_category_name: 610 (1.85%)
product_name_lenght: 610 (1.85%)
product_description_lenght: 610 (1.85%)
product_photos_qty: 610 (1.85%)
product_weight_g: 2 (0.01%)
product_length_cm: 2 (0.01%)
product_height_cm: 2 (0.01%)
product_width_cm: 2 (0.01%)
"""

products['product_category_name'] = products['product_category_name'].fillna('Unknown')
print(products.isnull().sum())

product_id                      0
product_category_name           0
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64


## Step 4 — Data types

The CSVs load everything as `object`/`float64`. Before the data can live in a typed SQL schema, timestamps must be real datetimes and integer-like counts should not be floats. Conversions use `errors='raise'` on purpose — if any value fails to parse, I want the pipeline to stop loudly rather than silently insert `NaT`.


In [12]:
for name, df in dfs.items():
    print(f"\n--- {name} ---")
    for col, dtype in df.dtypes.items():
        print(f"{col}: {dtype}")


--- customers ---
customer_id: str
customer_unique_id: str
customer_zip_code_prefix: int64
customer_city: str
customer_state: str

--- geolocation ---
geolocation_zip_code_prefix: int64
geolocation_lat: float64
geolocation_lng: float64
geolocation_city: str
geolocation_state: str

--- order_items ---
order_id: str
order_item_id: int64
product_id: str
seller_id: str
shipping_limit_date: str
price: float64
freight_value: float64

--- order_payments ---
order_id: str
payment_sequential: int64
payment_type: str
payment_installments: int64
payment_value: float64

--- order_reviews ---
review_id: str
order_id: str
review_score: int64
review_comment_title: str
review_comment_message: str
review_creation_date: str
review_answer_timestamp: str

--- orders ---
order_id: str
customer_id: str
order_status: str
order_purchase_timestamp: str
order_approved_at: str
order_delivered_carrier_date: str
order_delivered_customer_date: str
order_estimated_delivery_date: str

--- products ---
product_id: st

In [13]:
# Order items shipping_limit_date should be datetime instead of str
order_items['shipping_limit_date'] = pd.to_datetime(order_items['shipping_limit_date'], errors = 'raise')
print(order_items.dtypes)

order_id                          str
order_item_id                   int64
product_id                        str
seller_id                         str
shipping_limit_date    datetime64[us]
price                         float64
freight_value                 float64
dtype: object


In [14]:
# Order Reviews review_creation_date, review_answer_timestamp both should be datetime instead of str
cols = ['review_creation_date', 'review_answer_timestamp']

for col in cols:
    order_reviews[col] = pd.to_datetime(order_reviews[col], errors='raise')

print(order_reviews.dtypes)

review_id                             str
order_id                              str
review_score                        int64
review_comment_title                  str
review_comment_message                str
review_creation_date       datetime64[us]
review_answer_timestamp    datetime64[us]
dtype: object


In [15]:
# Orders order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date
# All should be datetime instead of str

cols = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

for col in cols:
    orders[col] = pd.to_datetime(orders[col], errors= 'raise')

print(orders.dtypes)

order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object


In [16]:
# Products
# Fix typos & change product_name_lenght, product_description_lenght, product_photos_qty to nullable Int64 since they only have whole numbers
products = products.rename(columns={
    'product_name_lenght': 'product_name_length',
    'product_description_lenght': 'product_description_length'
})

cols = ['product_name_length', 'product_description_length', 'product_photos_qty']
for col in cols:
    products[col] = products[col].astype('Int64')
print(products.dtypes)

product_id                        str
product_category_name             str
product_name_length             Int64
product_description_length      Int64
product_photos_qty              Int64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object


## Step 5 — Relational schema and load into PostgreSQL

Rather than letting `pandas.to_sql` auto-create untyped, constraint-free tables, the schema is declared explicitly with SQLAlchemy `Table` objects. The point is to make the database **enforce** the relationships the analysis will rely on:

- **Primary keys** on every dimension (`customers`, `products`, `sellers`, `geolocation`, `orders`)
- **Composite primary keys** where a single column can't be unique: `order_items` (`order_id` + `order_item_id`) and `order_payments` (`order_id` + `payment_sequential`)
- **`order_reviews` keyed on `order_id`** — valid only because of the Step 1 dedup decision
- **Foreign keys** from every child table to its parent, so a bad load fails loudly instead of silently producing orphans

This surfaced a referential-integrity issue during the initial load — see below.


In [17]:
from sqlalchemy import create_engine, text, MetaData, Table, Column, Integer, String, Float, DateTime, ForeignKey

engine = create_engine("postgresql+psycopg2://postgres:password@localhost:5432/olist")

metadata = MetaData()

customers_table = Table(
    "customers", metadata,
    Column("customer_id", String(32), primary_key=True),
    Column("customer_unique_id", String(32)),
    Column("customer_zip_code_prefix", Integer, ForeignKey("geolocation.geolocation_zip_code_prefix")),
    Column("customer_city", String),
    Column("customer_state", String)
)

geolocation_table = Table(
    "geolocation", metadata,
    Column("geolocation_zip_code_prefix", Integer, primary_key= True),
    Column("geolocation_lat", Float),
    Column("geolocation_lng", Float),
    Column("geolocation_city", String),
    Column("geolocation_state", String)
)

order_items_table = Table(
    "order_items", metadata,
    Column("order_id", String(32), ForeignKey("orders.order_id"),primary_key=True),
    Column("order_item_id", Integer, primary_key=True),
    Column("product_id", String(32), ForeignKey("products.product_id")),
    Column("seller_id", String(32), ForeignKey("sellers.seller_id")),
    Column("shipping_limit_date", DateTime),
    Column("price", Float),
    Column("freight_value", Float)
)

order_payments_table = Table(
    "order_payments", metadata,
    Column("order_id", String(32), ForeignKey("orders.order_id"),primary_key=True),
    Column("payment_sequential", Integer, primary_key=True),
    Column("payment_type", String),
    Column("payment_installments", Integer),
    Column("payment_value", Float)
)

order_reviews_table = Table(
    "order_reviews", metadata,
    Column("review_id", String(32)),
    Column("order_id", String(32), ForeignKey("orders.order_id"), primary_key=True),
    Column("review_score", Integer),
    Column("review_comment_title", String),
    Column("review_comment_message", String),
    Column("review_creation_date", DateTime),
    Column("review_answer_timestamp", DateTime)
)

orders_table = Table(
    "orders", metadata,
    Column("order_id", String(32), primary_key=True),
    Column("customer_id", String(32), ForeignKey("customers.customer_id")),
    Column("order_status", String),
    Column("order_purchase_timestamp", DateTime),
    Column("order_approved_at", DateTime),
    Column("order_delivered_carrier_date", DateTime),
    Column("order_delivered_customer_date", DateTime),
    Column("order_estimated_delivery_date", DateTime)
)

products_table = Table(
    "products", metadata,
    Column("product_id", String(32), primary_key=True),
    Column("product_category_name", String),
    Column("product_name_length", Integer),
    Column("product_description_length", Integer),
    Column("product_photos_qty", Integer),
    Column("product_weight_g", Integer),
    Column("product_length_cm", Integer),
    Column("product_height_cm", Integer),
    Column("product_width_cm", Integer)
)

sellers_table = Table(
    "sellers", metadata,
    Column("seller_id", String(32), primary_key=True),
    Column("seller_zip_code_prefix", Integer, ForeignKey("geolocation.geolocation_zip_code_prefix")),
    Column("seller_city", String),
    Column("seller_state", String)
)

translation_table = Table(
    "translation", metadata,
    Column("product_category_name", String),
    Column("product_category_name_english", String)
)

metadata.create_all(engine)

### Foreign-key violation at load: missing zip prefixes

The initial load failed with a foreign-key violation: `customers` and `sellers` reference zip prefixes that don't exist in the cleaned `geolocation` table. The first step is to quantify the problem before choosing a fix.


In [18]:
# Error in loading the DataFrame into sql due to the mishaped geolocation table unable to connect to customers table
# Quantifying the issue then decise
missing_zips = customers.loc[~customers['customer_zip_code_prefix'].isin(geolocation['geolocation_zip_code_prefix']), 'customer_zip_code_prefix'].unique()

missing_zips_sellers = sellers.loc[~sellers['seller_zip_code_prefix'].isin(geolocation['geolocation_zip_code_prefix']), 'seller_zip_code_prefix'].unique()
all_missing_zips = np.unique(np.concatenate([missing_zips, missing_zips_sellers]))

print(len(all_missing_zips))

162


162 zip prefixes are referenced but missing. Options: drop the FK constraint (gives up integrity), delete the affected customers/sellers (destroys real sales data), or insert **placeholder geolocation rows** (`lat`/`lng` = NA, city/state = `'Unknown'`). The placeholder option keeps the FK enforceable and keeps every order analyzable — the trade-off is that these 162 zips can't be mapped geographically, which is acceptable because geographic analysis happens at the *state* level (from `customers.customer_state`), not from lat/lng.


In [19]:
# There are 162 unique zip code in customers and sellers table that doesn't exist in geolocation table
# Hence the decision is to create a placeholder row in geolocation table

placeholder_rows = pd.DataFrame({
    'geolocation_zip_code_prefix': all_missing_zips,
    'geolocation_lat': pd.NA,
    'geolocation_lng': pd.NA,
    'geolocation_city': 'Unknown',
    'geolocation_state': 'Unknown'
})

geolocation = pd.concat([geolocation, placeholder_rows], ignore_index=True)
geolocation['geolocation_lat'] = geolocation['geolocation_lat'].astype('Float64')
geolocation['geolocation_lng'] = geolocation['geolocation_lng'].astype('Float64')


print(geolocation.shape)  # should be 19015 + len(all_missing_zips) = 19177
print(geolocation['geolocation_lat'].dtype)
print(geolocation['geolocation_lng'].dtype)

(19177, 5)
Float64
Float64


### Rebuild and load

To make the load repeatable, the database is wiped and rebuilt from scratch each run: drop every table regardless of FK dependency order (`CASCADE`), verify the schema is empty, then recreate all 9 tables from the `Table` definitions above.


In [20]:
# 1. Drop every table in the database, regardless of FK dependencies
with engine.connect() as conn:
    conn.execute(text("""
        DO $$ 
        DECLARE
            r RECORD;
        BEGIN
            FOR r IN (SELECT tablename FROM pg_tables WHERE schemaname = 'public') LOOP
                EXECUTE 'DROP TABLE IF EXISTS "' || r.tablename || '" CASCADE';
            END LOOP;
        END $$;
    """))
    conn.commit()

# 2. Confirm the database is actually empty now
with engine.connect() as conn:
    result = conn.execute(text("SELECT tablename FROM pg_tables WHERE schemaname = 'public'"))
    print(result.fetchall())  # should print []

# 3. Rebuild all 9 tables fresh from your Table() definitions
metadata.create_all(engine)

# 4. Confirm they're back, with correct structure
with engine.connect() as conn:
    result = conn.execute(text("SELECT tablename FROM pg_tables WHERE schemaname = 'public'"))
    print(result.fetchall())

[]
[('translation',), ('geolocation',), ('customers',), ('sellers',), ('orders',), ('order_items',), ('products',), ('order_payments',), ('order_reviews',)]


Load order matters with enforced FKs: parents before children. `geolocation` goes first (referenced by `customers`/`sellers`), then the remaining tables in dependency order.


In [21]:
geolocation.to_sql("geolocation", engine, if_exists="append", index=False)

with engine.connect() as conn:
    result = conn.execute(text("SELECT COUNT(*) FROM geolocation"))
    print(result.fetchone())

(19177,)


In [22]:
dfs_to_load = {
    "translation": translation,
    "customers": customers,
    "sellers": sellers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
    "order_payments": order_payments,
    "order_reviews": order_reviews,
}

for table_name, df in dfs_to_load.items():
    df.to_sql(table_name, engine, if_exists="append", index=False)
    print(f"Loaded {table_name}: {len(df)} rows")

Loaded translation: 71 rows
Loaded customers: 99441 rows
Loaded sellers: 3095 rows
Loaded products: 32951 rows
Loaded orders: 99441 rows
Loaded order_items: 112650 rows
Loaded order_payments: 103886 rows
Loaded order_reviews: 98673 rows


## Step 6 — Referential integrity checks

Final verification: no child rows pointing at nonexistent parents. All three orphan counts must be zero — and they are. The FK constraints guarantee this by construction; the direct checks confirm it.


In [23]:
checks = {
    "orphaned order_items (bad order_id)": "SELECT COUNT(*) FROM order_items WHERE order_id NOT IN (SELECT order_id FROM orders)",
    "orphaned order_payments": "SELECT COUNT(*) FROM order_payments WHERE order_id NOT IN (SELECT order_id FROM orders)",
    "orphaned order_reviews": "SELECT COUNT(*) FROM order_reviews WHERE order_id NOT IN (SELECT order_id FROM orders)",
}
with engine.connect() as conn:
    for label, sql in checks.items():
        print(label, conn.execute(text(sql)).fetchone())

orphaned order_items (bad order_id) (0,)
orphaned order_payments (0,)
orphaned order_reviews (0,)


## Where this goes next

The database built here powers the rest of the project:

- **`Olist-Analysis-Log.md`** — six SQL analyses (delivery status vs. reviews, sales trend, seller performance, regional delivery time, repeat customers, RFM segmentation), each logged with the query, finding, and caveats.
- **`olist.pbix`** — a 4-page Power BI dashboard (Overview, Delivery, Seller, Customer) built on exports from this database.

See `README.md` for the full project story.
